mono/metadata/w32process-unix.c

```diff
@@ -741,45 +728,28 @@ processes_cleanup (void)
 		}
 	}
 
-	/*
-	 * Remove processes which exited from the processes list.
-	 * We need to synchronize with the sigchld handler here, which runs
-	 * asynchronously. The handler requires that the processes list
-	 * remain valid.
-	 */
 	mono_os_mutex_lock (&processes_mutex);
 
-	for (process = processes; process; process = process->next) {
+	for (process = processes; process;) {
+		Process *next = process->next;
 		if (process->handle_count == 0 && process->freeable) {
 			/*
 			 * Unlink the entry.
-			 * This code can run parallel with the sigchld handler, but the
-			 * modifications it makes are safe.
 			 */
 			if (process == processes)
 				processes = process->next;
 			else
 				prev->next = process->next;
-			finished = g_slist_prepend (finished, process);
+
+			mono_os_sem_destroy (&process->exit_sem);
+			memset (process, 42, sizeof (Process));
```

Comment:
Well, considering the amount of bugs we had here, it might be a good idea to leave it in :)